gwcs/wcs.py

```diff
@@ -634,3 +641,235 @@ def fix_inputs(self, fixed):
+        U = (cdmat[1, 1] * undist_x - cdmat[0, 1] * undist_y) / det
+        V = (-cdmat[1, 0] * undist_x + cdmat[0, 0] * undist_y) / det
+        fit_inv_poly_u, fit_inv_poly_v, max_resid, rms = fit_2D_poly(degree,
+                                                            max_inv_error, max_inv_rms,
+                                                            U, V, u - U, v - V,
+                                                            verbose=verbose)
+        pdegree = fit_poly_x.degree
+        hdr['a_order'] = pdegree
+        hdr['b_order'] = pdegree
+        ipdegree = fit_inv_poly_u.degree
+        hdr['ap_order'] = ipdegree
+        hdr['bp_order'] = ipdegree
+        cd, sip_poly_x, sip_poly_y = reform_poly_coefficients(fit_poly_x, fit_poly_y)
+        store_2D_coefficients(hdr, sip_poly_x, 'A')
+        store_2D_coefficients(hdr, sip_poly_y, 'B')
+        store_2D_coefficients(hdr, fit_inv_poly_u, 'AP', keeplinear=True)
+        store_2D_coefficients(hdr, fit_inv_poly_v, 'BP', keeplinear=True)
+        hdr['cd1_1'] = cd[0][0]
+        hdr['cd1_2'] = cd[0][1]
+        hdr['cd2_1'] = cd[1][0]
+        hdr['cd2_2'] = cd[1][1]
+        hdr['sipmxerr'] = max_resid
+        hdr['siprms'] = rms
+        # # for test purposes, make sip equivalent in GWCS
+        # cdmat = np.array([[cd[0][0], cd[0][1]], [cd[1][0], cd[1][1]]])
+        # # cdmat = [[cd[0][0], cd[1][0]], [cd[0][1], cd[1][1]]]
+        # sip_poly_test_x = sip_poly_x.copy()
+        # sip_poly_test_y = sip_poly_y.copy()
+        # # sip_poly_test_x.c1_0 = 1
+        # # sip_poly_test_y.c0_1 = 1
+        # shift = (Shift(-crpix1) & Shift(-crpix2))
+        # aff = AffineTransformation2D(cdmat, [0,0])
+
+        # transform = (shift | Mapping([0,1,0,1]) |
+        #             (Mapping([0], n_inputs=2) + sip_poly_test_x) &
+        #             (Mapping([1], n_inputs=2) + sip_poly_test_y) |
+        #             aff |
+        #             Pix2Sky_TAN() | RotateNative2Celestial(crval1, crval2, 180))
+        # transform2 = ((Shift(-crpix1) & Shift(-crpix2)) | Mapping([0,1,0,1]) |
+        #               (fit_poly_x & fit_poly_y) | Pix2Sky_TAN() |
+        #               RotateNative2Celestial(crval1, crval2, 180))
+        return hdr
+
+
+def fit_2D_poly(degree, max_error, max_rms, u, v, x, y, verbose=False):
+    """
+    Fit a pair of ordinary 2D polynomial to the supplied inputs (u, v) and
+    outputs (x, y)
+    """
+    llsqfitter = LinearLSQFitter()
+    # The case of one pass with the specified polynomial degree
+    if degree is not None:
+        poly_x = Polynomial2D(degree=degree)
+        poly_y = Polynomial2D(degree=degree)
+        fit_poly_x = llsqfitter(poly_x, u, v, x)
+        fit_poly_y = llsqfitter(poly_y, u, v, y)
+        max_resid, rms = compute_distance_residual(x, y,
+                                                   fit_poly_x(u, v), fit_poly_y(u, v))
+    else:
+        cond = True
+        degree = 2
```

Comment:
It is not clear why minimum degree should be 2 and not 1. What if there are no "distortions"?

Also, this same code:
```python
         poly_x = Polynomial2D(degree=degree)
         poly_y = Polynomial2D(degree=degree)
         fit_poly_x = llsqfitter(poly_x, u, v, x)
         fit_poly_y = llsqfitter(poly_y, u, v, y)
         max_resid, rms = compute_distance_residual(x, y, fit_poly_x(u, v), fit_poly_y(u, v))
```
is used twice. I think this whole big `if` and the `while` loop could be replaced with something like:
```python
for degree in range(10) if degree is None else [degree]:  # or call loop variable 'deg'
             poly_x = Polynomial2D(degree=degree)
    poly_y = Polynomial2D(degree=degree)
    fit_poly_x = llsqfitter(poly_x, u, v, x)
    fit_poly_y = llsqfitter(poly_y, u, v, y)
    max_resid, rms = compute_distance_residual(x, y, fit_poly_x(u, v), fit_poly_y(u, v))
    if max_resid ...:
        break
else:
    raise RuntimeError("Unable to achieve required fit accuracy using SIP polynomials of degree < 10")

Reply:
If degree is 1 there is no SIP polynomial. I wonder if that special case should be in this. Hmmm.

I don't find that loop construction very clear, but I do have an alternative.

Reply:
Well, it avoids `cond` and doubling of the code and it simplifies the code when to raise an exception.

Reply:
> I wonder if that special case should be in this. Hmmm.

I think that users probably just want a **`FITS WCS`** that _potentially_ might have SIP terms. I do not think users would _insist_ on having SIP coefficients in the header even when they are 0 (and do not contribute to improved results over a CD matrix).